src/java.base/share/classes/javax/crypto/KEMSpi.java

```diff
@@ -0,0 +1,245 @@
+
+        /**
+         * Returns the size of the key encapsulation message.
+         *
+         * @return the size of the key encapsulation message as a finite non-negative integer
+         * @see KEM.Encapsulator#encapsulationSize()
+         */
+        int engineEncapsulationSize();
+    }
+
+    /**
+     * The KEM decapsulator implementation, generated by
+     * {@link #engineNewDecapsulator}.
+     *
+     * @see KEM.Decapsulator
+     */
+    interface DecapsulatorSpi {
+        /**
+         * The key decapsulation function.
+         * <p>
+         * An implementation must support the case where {@code from} is 0,
+         * {@code from} is the same as the output of {@code secretSize()},
+         * and {@code algorithm} is "Generic".
+         * <p>
+         * The caller of this method has already validated the parameters to
+         * ensure that neither {@code encapsulation} not {@code algorithm} is
+         * {@code null}, and the values of {@code from} and {@code to} are
+         * within the correct range. Therefore an implementation of this method
+         * does not to validate them.
+         *
+         * @param encapsulation the key encapsulation message from the sender
+         * @param from the initial index of the shared secret to be returned, inclusive
+         * @param to the final index of the shared secret to be returned, exclusive.
+         * @param algorithm the algorithm for the secret key returned
+         * @return a portion of the shared secret as a {@code SecretKey} with
+         *          the specified algorithm, must not be {@code null}
+         * @throws UnsupportedOperationException if the combination of
+         *          {@code from}, {@code to}, and {@code algorithm}
+         *          is not supported by the decapsulator.
+         * @throws DecapsulateException if an error occurs during the
+         *          decapsulation process
+         * @see KEM.Decapsulator#decapsulate(byte[], int, int, String)
+         */
+        SecretKey engineDecapsulate(byte[] encapsulation, int from, int to, String algorithm)
+                throws DecapsulateException;
+
+        /**
+         * Returns the size of the shared secret.
+         *
+         * @return the size of the shared secret as a finite non-negative integer
+         * @see KEM.Decapsulator#secretSize()
+         */
+        int engineSecretSize();
+
+        /**
+         * Returns the size of the key encapsulation message.
+         *
+         * @return the size of the key encapsulation message as a finite non-negative integer
+         * @see KEM.Decapsulator#encapsulationSize()
+         */
+        int engineEncapsulationSize();
+    }
+
+    /**
+     * Creates a KEM encapsulator.
+     * <p>
+     * The caller of this method has already validated the parameters to
+     * ensure that {@code pk} is not {@code null}. Therefore an implementation
+     * of this method does not to validate it.
```

Comment:
Same as above, the caller may not validate the parameters yet.  For example, the instance could be accessed just like what you do in the KEM implementation manner (use Provider method and Service look-up APIs), but without validating the parameters as you did.